public/app/plugins/datasource/dashboard/models.cue

```diff
@@ -0,0 +1,32 @@
+// Copyright 2021 Grafana Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package grafanaplugin
+
+import "github.com/grafana/thema"
+
+Datasource: thema.#Lineage & {
```

Comment:
This'll need to have both `Query` and a `DSOptions` lineages, like this: https://gist.github.com/sdboyer/4f79e228fd5c1e9924b970c8ba60510b#file-models-cue

This is the [Query](https://github.com/grafana/grafana/blob/main/pkg/framework/coremodel/slots.cue#L43) meta-schema (unrestricted), and this is the [DSOptions](https://github.com/grafana/grafana/blob/main/pkg/framework/coremodel/slots.cue#L49-L54) meta-schema. (Doing the same thing for datasource plugins as the [Panel meta-schema](https://github.com/grafana/grafana/blob/main/pkg/framework/coremodel/slots.cue#L30-L40) is doing for panel plugins right now)

Reply:
Great -- but lets be super careful with:
```
slots: DSOptions: {
  // Normal datasource configuration options.
  Options: {...}
  // Sensitive datasource configuration options that require encryption.
  SecureOptions: {...}
}
```
since this *REALLY* *REALLY* needs some help before pushing any new pattern.  
1. actual "options" are some subset of the common ones on https://github.com/grafana/grafana/blob/main/pkg/api/dtos/datasource.go#L19
2. anything in [jsonData](https://github.com/grafana/grafana/blob/main/pkg/api/dtos/datasource.go#L26)
3. secure things... although the types are open, I *think* it can only ever actually be `map[string]strings`, but we want to say what strings should exist

Reply:
just noticed these comments x-posted!

The only thing really well-settled at this point is the `Panel` slot. What i have there right now for the `Query` and `DSOptions` slots aren't finalized.

This is such a foundational part of the overall system that there isn't a functional way of expressing the immaturity of these slots beyond some comments...which i should add.

In general, in the future, we should expect that any slot definitions, once declared in the way that `Query` and `DSOptions` are today, are a public, finalized contract that must follow standard backwards compatibility rules (all new slot defs, or fields in an existing slot def, are optional). This will need enforcement in CI.

To individual points:

> 1. actual "options" are some subset of the common ones on https://github.com/grafana/grafana/blob/main/pkg/api/dtos/datasource.go#L19
> 2. anything in [jsonData](https://github.com/grafana/grafana/blob/main/pkg/api/dtos/datasource.go#L26)

The way this should work out is that the datasource coremodel - whenever we get to defining it - specifies the base struct to which you linked, and stuff plugins define in `DSOptions.Options` specifies the expected structure of the [`JsonData`](https://github.com/grafana/grafana/blob/5e4fd94413387e3f9692acc49aa5de9442bb2951/pkg/api/dtos/datasource.go#L26) field within that struct, specifically for instances of that struct representing a particular plugin.

In Go, that obviously means relying on some kind of loose for the actual generated `JsonData` field type in the datasource coremodel's Go representation - e.g. `interface{}` or `map[string]interface{}`. (Even with Go generics, we can't enforce it with the type system, because we a) can't statically enumerate all plugins and b) can't express the discriminated union to link the contents of `JsonData` to `Type`.) That's OK - type safety comes from Thema & CUE, not Go's type checker. The goal is just to make a Go type [to which all valid instances of the schema will be assignable](https://github.com/grafana/thema/blob/main/docs/invariants.md#go-assignability).

The exact approach we take will be the same as in generated dashboard types, and the same in grafana as in grok - i just have to dig in and pick an approach, which grok will force me to do presently.

> 3. secure things... although the types are open, I think it can only ever actually be `map[string]string`s, but we want to say what strings should exist

I've seen different strategies elsewhere on encrypted data - do we treat it as just a blob (`[]byte`), or as a keyed map of blobs (`map[string][]byte`)? In the current DTO struct, it looks like we're setting ourselves up for the latter [by just having field names](https://github.com/grafana/grafana/blob/5e4fd94413387e3f9692acc49aa5de9442bb2951/pkg/api/dtos/datasource.go#L27). For codegen simplicity that approach may need to shift a bit (`SecureJsonData map[string][]byte` instead), but we can figure that out.

But much of the motivation for separating these declarations from their literal usage in the datasource coremodel is to give us flexibility on where and when we use them. For encrypted data, that "when" is key, because we clearly need to receive encrypted data and keep it in encrypted form, but then also be able to unserialize it into the appropriate decrypted form. To my mind, that inevitably means generating (at least) two types - one for an encrypted form, one for a decrypted form.

I haven't dug in on how our crypt cycle actually works, so i don't have a strong take yet on the right codegen pattern. But i am quite certain that, by letting the plugin author declare `DSOptions.Options` and `DSOptions.SecureOptions` independently, we're letting them unambiguously express their intent as to what's encrypted and what's not. So long as the plugin author's intent is unambiguous, we'll be able to map that intent onto whatever set of types we thinks works best.


I've seen encryption strategies where the whole field 

I suspect that